### PR TITLE
copy vhd into the testing location when source vhd location is different

### PR DESCRIPTION
**Pull request by LiliDeng**
just a draft, 3 things need to be fixed.
- when reuse rg, this logic still be executed, need fix.
- need shorten the copied vhd, currently it is too long.
- also if the vhd has been copied, if the sas url is different, it will still be copied again.

### CODE REVIEW

**Comment by squirrelsc** [comment]
Can you provide the reason of this change in the commit description?

**Comment by squirrelsc** [comment]
> * also if the vhd has been copied, if the sas url is different, it will still be copied again.
     Is it expected behavior, or something you need to fix?

File: lisa/sut_orchestrator/azure/common.py [hunk 1]
```diff
@@ -488,13 +495,38 @@ def get_storage_credential(
     return {"account_name": account_name, "account_key": key.value}
 
 
-def get_or_create_storage_container(
+def generate_sas(
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/common.py]: `generate_sas_token` as the returned value name.

File: lisa/sut_orchestrator/azure/platform_.py [hunk 2]
```diff
@@ -2016,8 +2050,17 @@ def _get_deployable_vhd_path(
         """
         matches = SAS_URL_PATTERN.match(vhd_path)
         if not matches:
-            return vhd_path
-        log.debug("found the vhd is a sas url, it may need to be copied.")
+            vhd_path = self._check_and_generate_sas_url(vhd_path, location, log)
+            matches = SAS_URL_PATTERN.match(vhd_path)
+            if not matches:
+                return vhd_path
+            log.debug(
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/platform_.py]: Why not direct copy without sas token?

File: lisa/sut_orchestrator/azure/platform_.py [hunk 3]
```diff
@@ -2084,6 +2127,26 @@ def _get_deployable_vhd_path(
 
         return full_vhd_path
 
+    def _get_storage_details(self, vhd_path: str) -> Any:
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/platform_.py]: `_get_vhd_details`?

File: lisa/sut_orchestrator/azure/platform_.py (hunk 2 above)
Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/platform_.py]: I tired, but didn't work, the exception is 
```
azure.core.exceptions.ClientAuthenticationError: Server failed to authenticate the request. Please refer to the information in the www-authenticate header.
RequestId:ccc36585-601e-000e-2959-a6a9db000000
Time:2022-08-02T10:17:54.9250330Z
ErrorCode:CannotVerifyCopySource
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>CannotVerifyCopySource</Code><Message>Server failed to authenticate the request. Please refer to the information in the www-authenticate header.
RequestId:ccc36585-601e-000e-2959-a6a9db000000
Time:2022-08-02T10:17:54.9250330Z</Message></Error>
```

Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/platform_.py]: probably I need find out another way, for this method start_copy_from_url, it needs to be ` The source blob must either be public or must be authenticated via a shared access signature. `

image

File: lisa/sut_orchestrator/azure/common.py (hunk 1 above)
Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/common.py]: updated

File: lisa/sut_orchestrator/azure/platform_.py (hunk 3 above)
Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/platform_.py]: updated

**Comment by LiliDeng** [comment]
@squirrelsc during my testing, I found if set concurrency as 2, then there are 2 sas url generated, vhd will be copied twice, can this design be accepted?

File: lisa/sut_orchestrator/azure/platform_.py [hunk 4]
```diff
@@ -1937,8 +1969,17 @@ def _get_deployable_vhd_path(
         """
         matches = SAS_URL_PATTERN.match(vhd_path)
         if not matches:
-            return vhd_path
-        log.debug("found the vhd is a sas url, it may need to be copied.")
+            vhd_path = self._generate_sas_token(vhd_path, location, log)
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/platform_.py]: This should be split into two method. It doesn't need to check if it's sas token again. 1) Check if it's in the same location. If yes, return the vhd path. 2) Generate sas token, and go with next logics.

Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/platform_.py]: updated

**Review by squirrelsc — APPROVED** [review]

**Pull request merged**